### PR TITLE
Support Kafka client 2.1.0+  Closes #5730

### LINKED ISSUE #5730: Agent plugin for Kafka 2.1.0+
kafka client 2.3.1

image

### PR DESCRIPTION
**Pull request by hsoftxl**
### Support Kafka client 2.1.0+

- [x] Add a test case for the new plugin, refer to [the doc](https://github.com/apache/skywalking/blob/master/docs/en/guides/Plugin-test.md)
- [x] Add a component id in [the component-libraries.yml](https://github.com/apache/skywalking/blob/master/oap-server/server-bootstrap/src/main/resources/component-libraries.yml). **NO NEED**
- [x] Add a logo in [the UI repo](https://github.com/apache/skywalking-rocketbot-ui/tree/master/src/views/components/topology/assets). **NO NEED**
- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #5730.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/c2141978d1039375598a32418b75161a78da22c1/CHANGES.md).

### CODE REVIEW

**Comment by kezhenxu94** [comment]
Hi @hsoftxl , thanks for you contribution, can you please follow the pull request template(I edited that for you) and check the items so that we can start reviewing, thanks.

**Comment by hsoftxl** [comment]
testcase 
org.apache.skywalking.apm.plugin.kafka.ProducerConstructorPropertiesInterceptor

added change log

**Comment by hsoftxl** [comment]
@kezhenxu94

File: apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/KafkaProducerPropertiesInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    public static final String CONSTRUCTOR_INTERCEPTOR_CLASS = "org.apache.skywalking.apm.plugin.kafka.ProducerConstructorPropertiesInterceptor";
+    public static final String CONSTRUCTOR_INTERCEPTOR_FLAG = "java.util.Properties";
+
+    @Override
+    public ConstructorInterceptPoint[] getConstructorsInterceptPoints() {
+        return new ConstructorInterceptPoint[]{
+                new ConstructorInterceptPoint() {
+                    @Override
+                    public ElementMatcher<MethodDescription> getConstructorMatcher() {
+                        return takesArgumentWithType(0, CONSTRUCTOR_INTERCEPTOR_FLAG);
```
Comment (kezhenxu94) [inline apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/KafkaProducerPropertiesInstrumentation.java:45]: I checked the `KafkaProducer` class in the version 2.3.1 (that you mentioned in the issue), and found that this constructor should be instrumented by `KafkaProducerMapInstrumentation`, this seems to be duplicated. Can you elaborate more? Thanks 

image

Comment (hsoftxl) [reply apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/KafkaProducerPropertiesInstrumentation.java:45]: KafkaProducerMapInstrumentation  
which version add this class

Comment (hsoftxl) [reply apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/KafkaProducerPropertiesInstrumentation.java:45]: i use 8.1.0 , can not support kafka2.3.1

Comment (kezhenxu94) [reply apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/KafkaProducerPropertiesInstrumentation.java:45]: > KafkaProducerMapInstrumentation. which version add this class

It's there for a long time.

> i use 8.1.0 , can not support kafka2.3.1
 
So did it work after your changes in this PR?

Comment (hsoftxl) [reply apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/KafkaProducerPropertiesInstrumentation.java:45]: yes

Comment (kezhenxu94) [reply apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/KafkaProducerPropertiesInstrumentation.java:45]: @hsoftxl do you use a `List` as `bootstrap.servers` in the properties?

Comment (kezhenxu94) [reply apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/KafkaProducerPropertiesInstrumentation.java:45]: I've rechecked 8.1.0, it should intercept the constructor you mentioned in this PR, and I believe the instrumentation you added in this PR is duplicated.

If you're using a `List` as `bootstrap.servers` in the properties, it has problem in 8.1.0 but if you rebuild the plugin from the `master` branch (with or without your changes in this PR), it should work, can you confirm?

Comment (hsoftxl) [reply apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/KafkaProducerPropertiesInstrumentation.java:45]: image
my properies

i rebuild the plugin from master, when i rebuild the plugin whitout my change , it has exception  , list / string case exception,  i will test one more time.

Comment (hsoftxl) [reply apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/KafkaProducerPropertiesInstrumentation.java:45]: image
my properies

i rebuild the plugin from master, when i rebuild the plugin whitout my change , it has exception  , list / string case exception,  i will test one more time.

image

Comment (hsoftxl) [reply apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/KafkaProducerPropertiesInstrumentation.java:45]: @kezhenxu94 mybey  the first time id rebulid the plugin is not base on master.
i will check again

Comment (kezhenxu94) [reply apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/KafkaProducerPropertiesInstrumentation.java:45]: image

Make sure you have the latest codes, the master branch have this type detection, but from the exception log, seems it's not the case

Comment (hsoftxl) [reply apm-sniffer/apm-sdk-plugin/kafka-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/define/KafkaProducerPropertiesInstrumentation.java:45]: the master code is well,       i builded the plugin  from 8.1.0 tag

**Pull request closed without merging**